jsonschema-core.md

```diff
@@ -912,10 +912,11 @@ the case of a network-addressable URL, a schema need not be downloadable from
 its canonical IRI.
 
 If present, the value for this keyword MUST be a string, and MUST represent a
-valid [IRI-reference](#rfc3987). This IRI-reference SHOULD be normalized, and
-MUST resolve to an [absolute-IRI](#rfc3987) (without a fragment).
+valid [IRI reference](#rfc3987). This IRI reference SHOULD be normalized per RFC
+3987, section 5.3, and MUST resolve to an [absolute IRI](#rfc3987) (without a
+fragment).
```

Comment:
I'll have a think about this.  I don't disagree, but I'll need to figure out where to put such a requirement.  This section does warrant a note to authors that implementations will be performing this normalization; probably a link to wherever this ends up.

Reply:
Leaving it in place and changing to

> If present, the value for this keyword MUST be a string, and MUST represent a valid [IRI reference](#rfc3987). When processing this IRI reference, implementations SHOULD normalize it per RFC 3987, section 5.3, so that it resolves to an [absolute IRI](#rfc3987) (without a fragment).

seems to work.  It puts the requirement on the implementation, but it also lets the user (if they're even reading this) know what they can expect from the implementation.

I do fear, though, that this gives license to users to put IRIs _with_ fragments in `$id`.  (It's almost like the output needs the ability to report a warning as well as an error. "Hey, uh, there's a fragment on that IRI, guy. You sure you want to do that? Probably don't.")

Reply:
I don't think leaving it in place works because `$id` isn't the only way a schema might be assigned an IRI. For example, implementations often also allow users to assign IRIs to schemas when they're registered. The normalization requirement should also apply to those user provided IRIs and therefore needs to be somewhere where it doesn't only apply to `$id`.

One place I think it could work is in "Base IRI, Anchors, and Dereferencing" above. Here's a suggestion.

> To differentiate between schemas in a vast ecosystem, schemas are identified by [absolute IRIs](#rfc3987) (without a fragment).
>
> Schemas can embed references to other schemas by specifying their IRI. When implementations dereference these references, they SHOULD use the normalization procedures defined in RFC 3987, section 5.3 when comparing URIs.
>
> Several keywords can accept an [IRI reference](#rfc3987). The schema's identifier serves as the base IRI for resolving relative references.

Then I'd remove everything about normalization and fragments from this paragraph. If we want, we could include an informal warning that fragments are ignored because schemas are identified by absolute IRIs, but it shouldn't technically be necessary.

Reply:
We _do_ need information on valid values for `$id`, though.  I feel that's what at least the first sentence does.  I'll play with it.

Reply:
Right, I was only suggesting removing the part about normalization and fragments. The part about it being an IRI reference needs to stay.

Reply:
I've moved the normalization up to the previous section as suggested.  I had to play with the wording a bit.